src/json/json.cc

```diff
@@ -0,0 +1,3232 @@
+    if (command == nullptr) {
+        ValkeyModule_Log(ctx, "warning", "Failed to get command '%s'", name);
+        return false;
+    }
+    ValkeyModuleCommandInfo info;
+    memset(&info, 0, sizeof(info));
+    info.version = VALKEYMODULE_COMMAND_INFO_VERSION;
+    info.arity = arity;
+
+    // We only need one key_spec entry, but key_specs are sent as a null-entry terminated array,
+    // so we leave a second value filled with 0s
+    ValkeyModuleCommandKeySpec cmdKeySpec[2];
+    memset(cmdKeySpec, 0, sizeof(cmdKeySpec));
+
+    cmdKeySpec[0].flags = keyspec_flags;
+    cmdKeySpec[0].begin_search_type = VALKEYMODULE_KSPEC_BS_INDEX;
+    cmdKeySpec[0].bs.index = {bs_index};
+    cmdKeySpec[0].find_keys_type = VALKEYMODULE_KSPEC_FK_RANGE;
+    cmdKeySpec[0].fk.range = {std::get<0>(key_range), std::get<1>(key_range), std::get<2>(key_range)};
+
+    info.key_specs = &cmdKeySpec[0];
+
+    if (ValkeyModule_SetCommandInfo(command, &info) == VALKEYMODULE_ERR) {
+        ValkeyModule_Log(ctx, "warning", "Failed to set command info for %s", name);
+        return false;
+    }
+    return true;
+}
+
+/* ================================ Module OnLoad ============================= */
+
+extern "C" int ValkeyModule_OnLoad(ValkeyModuleCtx *ctx) {
+    // Register the module
+    if (ValkeyModule_Init(ctx, MODULE_NAME, MODULE_VERSION, VALKEYMODULE_APIVER_1) == VALKEYMODULE_ERR) {
+        ValkeyModule_Log(ctx, "warning", "Failed to initialize module %s version %d", MODULE_NAME, MODULE_VERSION);
+        return VALKEYMODULE_ERR;
+    }
+
+    // Register module type callbacks
+    ValkeyModuleTypeMethods type_methods;
+    memset(&type_methods, 0, sizeof(ValkeyModuleTypeMethods));
+    type_methods.version = VALKEYMODULE_TYPE_METHOD_VERSION;
+    type_methods.rdb_load = DocumentType_RdbLoad;
+    type_methods.rdb_save = DocumentType_RdbSave;
+    type_methods.copy = DocumentType_Copy;
+    type_methods.aof_rewrite = DocumentType_AofRewrite;
+    type_methods.mem_usage = DocumentType_MemUsage;
+    type_methods.free = DocumentType_Free;
+    type_methods.digest = DocumentType_Digest;
+    type_methods.defrag = DocumentType_Defrag;
+
+
+    // Create module type
+    DocumentType = ValkeyModule_CreateDataType(ctx, DOCUMENT_TYPE_NAME,
+                                              DOCUMENT_TYPE_ENCODING_VERSION, &type_methods);
+    if (DocumentType == nullptr) {
+        ValkeyModule_Log(ctx, "warning", "Failed to create data type %s encver %d",
+                        DOCUMENT_TYPE_NAME, DOCUMENT_TYPE_ENCODING_VERSION);
+        return VALKEYMODULE_ERR;
+    }
+
+    /*
+     * Now create the stub datatypes for search
+     */
+    if (!install_stub(ctx, "scdtype00", SCDTYPE_ENCVER, scdtype_aux_load)) return VALKEYMODULE_ERR;
+    if (!install_stub(ctx, "GEARS_DT0", GEARSDT_ENCVER, gearsdt_aux_load)) return VALKEYMODULE_ERR;
+    if (!install_stub(ctx, "GEAR_REQ0", GEARSRQ_ENCVER, gearsrq_aux_load)) return VALKEYMODULE_ERR;
```

Comment:
Apparently we also load 3 custom module types here. I understand this to be for compatibility with stack, but do we want to have this in this module?

Reply:
Intersting. At least, this should have be added to the rfc.

Reply:
Yes, that's for RDB compatibility with RedisJSON. If we don't care about them any more, we can remote it.

Reply:
I think we do care about compatibility with RedisJSON. We want to make it easy to migrate to Valkey.

I just wish it is mentioned in the RFC, which already has a lot of details.

Reply:
These aren't for json compatibility though, they list gears and search types.

Reply:
Which are partially used by the JSON module or vice versa? Or the JSON module provides a minimal subset of them or something? I don't know this. :laughing:

I guess the only reason to add them is to support RDB streams containing module types from these modules. Otherwise, delete it.

Reply:
yeah these aux data types seem to just play a supporting role so far. can we register a generic stub instead and just log a message that "xxx is skipped" and move on? there could be more aux types added in the future in rejson so it will be good to have that visibility if community members report any issue.

```c
install_stub(ctx, "*", GENERIC_ENCVER, generic_aux_load);
```

Reply:
> These aren't for json compatibility though, they list gears and search types.

When testing compatibility with ReJSON, we use docker image "docker run -p 6389:6379 redis/redis-stack:latest", which loads a bunch of modules including search, graph, etc. Although our test only creates json keys, the generated RDB contains a bunch of aux types like scdtype00, gear*, etc. I suspect these aux types are from RedisSearch module, but not 100% sure because I never looked at their source code. (NOTE: It also indicates that their modules do not use aux_save2, which would not save aux types that have no corresponding data.)

To be able to load the RDB, we create stub structs for these aux types, read the data and ignored them. So, these are for JSON compatibility test.

Reply:
You are right @joehu21. These aux data types are indeed needed. The main purpose of these handlers is actually to skip these aux types.

Reply:
If we ever want another module that provides compatibilty with redis search or gears modules, then these name is blocked by ValkeyJSON module so the other modules can't be loaded? That's a potential future problem.

Reply:
Yeah, I think we need a better overall strategy here instead of install random types and ignoring the data from them.

Reply:
https://github.com/valkey-io/valkeyJSON/issues/9